tree/tree/src/TTree.cxx

```diff
@@ -6021,9 +6021,7 @@ TLeaf* TTree::GetLeafImpl(const char* branchname, const char *leafname)
    }
    TIter nextl(GetListOfLeaves());
    while ((leaf = (TLeaf*)nextl())) {
-      if (!strcmp(leaf->GetFullName(),leafname))
-         return leaf;
-      if (strcmp(leaf->GetName(),leafname))
+      if (strcmp(leaf->GetFullName(), leafname) && strcmp(leaf->GetName(), leafname))
```

Comment:
If the `leafname` argument matches the return value of `GetFullName` then (if all is as intended), it does by definition match the branch name (which is included in the `full name`).  However indeed the user could be calling 
```
GetLeaf("somebranchname", "someotherbranchname.leafname")
```
.. right and this might somehow be valid / intentional.  To support this we might need:
```suggestion
      if (!strcmp(leaf->GetFullName(), leafname) {
          if (!strcmp(leaf->GetBranch()->GetName(), branchname)
             return;
          else
             continue;
      }
      if (strcmp(leaf->GetName(), leafname))
```

Reply:
> If the leafname argument matches the return value of GetFullName then (if all is as intended), it does by definition match the branch name

This is not true e.g. for `GetLeaf("notavalidbranch", "branchelementname")`. This patch fixes precisely this case: we were returning a non-null pointer as if `GetLeaf("branchelementname")` was called, and now we correctly return a null pointer.

Why is the patch problematic?

Reply:
Because the "shortcut" (that includes more precise match):
```
if (!strcmp(leaf->GetFullName(),leafname))
   return leaf;
```
is removed and thus we are returning to the prior incorrect and/or slow match.

I.e. ROOT-10942 is likely to return in some shape or form.

Reply:
I.e. I think that the change I propose should keep the shortcut and fix your use case.

Reply:
ok, thanks! i'll be back on this after the holidays :)

Reply:
Slightly reworked version (fixed missing brackets, added check for `!branchname`, added comments):

```suggestion
      if (!strcmp(leaf->GetFullName(), leafname)) {
         // if branchname is also a match, it's the correct leaf, else keep looking
         if (!branchname || !strcmp(leaf->GetBranch()->GetName(), branchname))
            return leaf;
         else
            continue;
      }
      if (strcmp(leaf->GetName(), leafname)) // no match
```

However: why is it ok to just check `leaf->GetBranch()->GetName()` against `branchname` here, while a few lines below, in order to check whether the branch name is a match, we need a much more complex logic (including parent branches etc.)?

Reply:
humm ... isn't it because it is a "shortcut"/"easy match" and that the complex check should only be implemented once?

Reply:
humm ... except for the "continue" being in the way ...

```suggestion
      if (!strcmp(leaf->GetFullName(), leafname)) {
         // if branchname is also a match, it's the correct leaf, else keep looking
         if (!branchname || !strcmp(leaf->GetBranch()->GetName(), branchname))
            return leaf;
      }
      if (strcmp(leaf->GetName(), leafname)) // no match
```

Reply:
Should be resolved as agreed after discussion in the latest commits